src/context.rs

```diff
@@ -233,11 +271,21 @@ impl Context {
             if let Ok(mut stream) =
                 TcpStream::connect(format!("{}:{}", socket_addr.ip(), socket_addr.port() + 10))
             {
-                let signal = bincode::serialize(&Signal::ShutDown).unwrap();
-                let mut message = ::capnp::message::Builder::new_default();
-                let mut task_data = message.init_root::<serialized_data::Builder>();
-                task_data.set_msg(&signal);
-                serialize_packed::write_message(&mut stream, &message).unwrap();
+                let buf = {
+                    let signal = bincode::serialize(&Signal::ShutDownGraceful).unwrap();
+                    let mut message = ::capnp::message::Builder::new_default();
+                    let mut task_data = message.init_root::<serialized_data::Builder>();
+                    task_data.set_msg(&signal);
+                    let mut buf = Vec::with_capacity(signal.len() + 64);
```

Comment:
what is the point of this?

Reply:
Because I required to get the exact length of the sent serialized capnp message. I don't know if there is a method to get it from `task_data` here tbh, was a quick copy paste, if there is a way is not required to insert it into a vec first.

Long explanation:
First is required to send the length of the upcoming message, and then send the actual message. On the receiving side you read the exact amount of incoming bytes. Doing otherwise (using read_to_end e.g.) is VERY problematic with non-blocking communication, as it will deadlock the open stream more often than not, could argue that here is not that necessary as the other side will be shutting down but this is a future open footgun and it really does not matter much so better safe than sorry.

Reply:
What would be ideal here (and elsewhere) though is to use the capnp async bufreader, I have to give it an other try to see if I can make it work somehow.

Reply:
If you use, capnp bufreader then you won't require to know the exact length of the message and you won't need read_to_end.

Reply:
This is one of the reasons I am thinking hard about async. It makes code difficult to reason about and the fact that so many crates are not runtime agnostic makes me even warier.  Sure we are dealing with lot of downloads from external systems like cloud storage and DB which would hugely benefit from async code. But we are not going to spawn thousands of tasks. Think about this, let's say we are going to use one big partitioned parquet folder from s3. First, we are going to get metadata of folders and then files and guess the number of partitions. Active number of partitions will always be limited by CPU cores. So we will spawn just 8-100 tasks only.  It seems reasonable to go with good old threads to spawn tasks here. And reasoning about them is much easier than async. However, I do see benefits from the async code and it can prove out to be a game-changer when dealing with a lot of small files. But just not sure it's worth the effort. We can spend our valuable time in developing ecosystem behind the crate itself. I am occupied with experiments in data frames. But here onwards, I will try to commit frequently. My first priority, I will finish essential interfaces for filesystem so that we can start implementing filesystems which would make this crate much more useful immediately to end-users. It is a huge work which can be divided into tasks very easily so that multiple people can work.

Reply:
I think we can maintain an hybrid approach where it fits, I have found how to work around it so we can maintain a non pure async neither pure sync. But yes makes things harder for sure (if anything because a lot of features seem on progress, not polished, etc.).

I am sure most things have workarounds is just a matter of finding out what is lacking and if there is a workaround, so mostly lack of knowledge by me with the async ecosystem. I don't mind keep experimenting (although it gets extremely annoying sometimes :D) until it's working and only commit then gradually.

Agree with your assessment though in many cases it may not be worth it...